### PR TITLE
🚧Immaturable skip list implement.

### PR DESCRIPTION
**Pull request by LebranceBW**
Some basic fn, some test cases but document hasn't been written.
 I'm not confident about the code. could you take a review 🔖 😄?

### CODE REVIEW

File: src/collections/skip_list/mod.rs [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+
+    #[test]
+    fn test_get_mut() {
+        let mut list = create_testee();
+        let mut ele = list.get_mut(&"a".to_owned()).unwrap();
+        *ele = 114514;
+        assert_eq!(*list.get_mut(&"a".to_owned()).unwrap(), 114514);
+    }
+
+    #[test]
```
Comment (weihanglo) [inline src/collections/skip_list/mod.rs:416]: Thought this should be a [`#[bench]`](https://doc.rust-lang.org/nightly/unstable-book/library-features/test.html) test but not unit test.

File: src/collections/skip_list/mod.rs [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+        write!(f, "SkipList:\n {:?}", self.head)
+    }
+}
+
+impl<K, V> Drop for SkipList<K, V>
+where
+    K: Default,
+    V: Default,
+{
+    fn drop(&mut self) {
```
Comment (weihanglo) [inline src/collections/skip_list/mod.rs:223]: Have only skimmed unsafe part but found that something drops twice unexpectly. You can add a test to check dropping count.

```rust
#[test]
fn drop() {
    static mut DROPS: u32 = 0;

    #[derive(Default)]
    struct S;

    impl Drop for S {
        fn drop(&mut self) {
            unsafe {
                DROPS += 1;
            }
        }
    }

    let count = 10;
    let mut d = SkipList::new();
    (0..count).for_each(|k| d.insert(k, S));

    core::mem::drop(d);

    unsafe {
        assert_eq!(DROPS, count);
    }
}
```

**Review by weihanglo — COMMENTED** [review]
Overall looks great! Just need to be more ergonomic As a library user, the `default` constraint is way too limited. Also, the code could be more idiomatic, Rust provides lots of handy feature you can use without hardcoded or explicitly denoting.

In additional, have you ever tried implementing this with `Rc`? I think it might work to avoid any single line of unsafe Rust.

**Comment by LebranceBW** [comment]
@weihanglo Thank you very much for taking the time to review my code, 😃  and I'm sorry for my poor simple mistakes.

whereas, I have tried the implement the skiplist with `Rc`, but I'm trapped by the implement of `iter` just like the [book](https://rust-unofficial.github.io/too-many-lists/fourth-iteration.html). for example:
```Rust
struct SkipNode<K, V> {
    entry: Entry<K, V>,
    next_by_height: [Option<Rc<RefCell<SkipNode>>>; MAX_HEIGHT],
}
impl<K, V> SkipNode<K, V> {
   pub fn iter() -> Iter<K, V> {
     ...//omittied
  }
}
pub struct Iter<'a, K, V>
{
    ptr: Option<Rc<RefCell<SkipNode<K, V>>>, // 1
    _marker: marker::PhantomData<&'a K>,
}
impl<'a, K, V: 'a> Iterator for Iter<'a, K, V>
{
    type Item =  std::cell::Ref('a, (K, V)); //2
    fn next(&mut self) -> Option<Self::Item> {
        self.ptr.map(|node| node.as_ref().borrow()); //3
    }
}
```
Because the RefCell holds the SkipNode, and the `borrow()` fn return a `Ref`. The code above will complain at 2 about return the temporary variable. I don't know how to overcome. Could you provide a solution? or tell me some other code just like this?

File: src/collections/skip_list/mod.rs [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+            0xFBFFFFFDu32,
+            0xFEFFFFFCu32,
+            0xFFBFFFFBu32,
+            0xFFEFFFFAu32,
+            0xFFFBFFF9u32,
+            0xFFFEFFF8u32,
+            0xFFFFBFF7u32,
+            0xFFFFEFF6u32,
+            0xFFFFFBF5u32,
+        ];
```
[the remark this answers is not in the document]
Comment (LebranceBW) [reply src/collections/skip_list/mod.rs:46]: Not picky, magic number is harder to understand compared to code.

File: src/collections/skip_list/mod.rs [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+    //         unsafe {
+    //             while let Some(next_node) = (*start_point).next_by_height[level] {
+    //                 if (*next_node).entry.key == *key {
+
+    //                 }
+    //             }
+    //         }
+    //     }
+    //     None;
+    // }
```
[the remark this answers is not in the document]
Comment (LebranceBW) [reply src/collections/skip_list/mod.rs:156]: I forget to complete it. 😢

File: src/collections/skip_list/mod.rs [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+    pub fn values(&self) -> Box<dyn Iterator<Item = &V> + '_> {
+        Box::new(self.iter().map(|(_, value)| value)) as Box<dyn Iterator<Item = &V>>
+    }
+    /// Creates an anonymous iterator yields the values in range [key1, key2).
+    pub fn range(&self, key1: K, key2: K) -> Box<dyn Iterator<Item = (&K, &V)> + '_> {
+        Box::new(
+            self.iter()
+                .filter(move |(key, _)| (*key >= &key1) && (*key < &key2)),
+        )
+    }
```
[the remark this answers is not in the document]
Comment (LebranceBW) [reply src/collections/skip_list/mod.rs:187]: 😄 new knowledge！

File: src/collections/skip_list/mod.rs [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+        while level < RAND_TABLE.len() && rand_num > RAND_TABLE[level] {
+            level += 1;
+        }
+        level
+    }
+    /// Creates an empty skip list.
+    pub fn new() -> Self {
+        SkipList {
+            head: Box::into_raw(Box::new(SkipNode {
+                entry: Default::default(),
```
[the remark this answers is not in the document]
Comment (LebranceBW) [reply src/collections/skip_list/mod.rs:58]: The `Default` bound servers to dummy head node, which is unnecessary. I will remove.

File: src/collections/skip_list/mod.rs (hunk 2 above)
Comment (LebranceBW) [reply src/collections/skip_list/mod.rs:223]: Memory error is hard for me noob, I will try turn to the Safe Rust.

File: src/collections/skip_list/mod.rs [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+{
+    head: *mut SkipNode<K, V>,
+    rng: rand::rngs::ThreadRng,
+    len: usize,
+}
+// ANCHOR_END: struct
+
+impl<K: Ord + Default, V: Default> SkipList<K, V> {
+    fn random_height(&mut self) -> usize {
+        // 超几何分布打的表，参考leveldb
```
[the remark this answers is not in the document]
Comment (LebranceBW) [reply src/collections/skip_list/mod.rs:31]: yeah, written mistake.

File: src/collections/skip_list/mod.rs [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+    /// Creates an anonymous iterator yields the key in specifi order.
+    pub fn keys(&self) -> Box<dyn Iterator<Item = &K> + '_> {
+        Box::new(self.iter().map(|(key, _)| key)) as Box<dyn Iterator<Item = &K>>
+    }
+    /// Creates an anonymous iterator yields the values in specifi order.
+    pub fn values(&self) -> Box<dyn Iterator<Item = &V> + '_> {
+        Box::new(self.iter().map(|(_, value)| value)) as Box<dyn Iterator<Item = &V>>
+    }
+    /// Creates an anonymous iterator yields the values in range [key1, key2).
+    pub fn range(&self, key1: K, key2: K) -> Box<dyn Iterator<Item = (&K, &V)> + '_> {
```
[the remark this answers is not in the document]
Comment (LebranceBW) [reply src/collections/skip_list/mod.rs:182]: Got it

**Pull request closed without merging**